### DISCUSSION TITLE
Optimizer does not optimize all hyperparameters.

### DISCUSSION THREAD

**Discussion by robl01** [cornellius-gp/gpytorch#2015] [Q&A]
Hello dear all,

I am quite new to GPyTorch and I am figuring out, how it works properly. At the moment I have troubles with my optimizer. Following the structure of the SimpleRegression Tutorial, I build my Gaussian Process and so far, it seems to optimize the loss but not other parameters. 

I have no clue, where the problem comes from. Thus, I think providing you the whole code is probably the fastest way. ( I work in a notebook environment :) ) 
```
import numpy as np 
import torch 
import gpytorch
import matplotlib.pyplot as plt
from matplotlib import cm
import matplotlib.colors as mcolors
import statistics # from Python library
from scipy.stats import  norm


class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self,x_train, y_train, likelihood):
        super(ExactGPModel,self).__init__(x_train,y_train,likelihood)
        self.mean_module = gpytorch.means.ConstantMean() # prior mean
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel(ard_num_dims=1))

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return new_MultivariateNormal(mean_x,covar_x)

class new_MultivariateNormal(gpytorch.distributions.MultivariateNormal):
        def confidence_region(self):
            """
        Returns 2 standard deviations above and below the mean.

        :rtype: (torch.Tensor, torch.Tensor)
        :return: pair of tensors of size (b x d) or (d), where
            b is the batch size and d is the dimensionality of the random
            variable. The first (second) Tensor is the lower (upper) end of
            the confidence region.
            """

            std2 = self.stddev
            mean = self.mean
            return mean.sub(std2), mean.add(std2)



        def std_dev(self):
            """
        Returns standard deviation.

        :rtype: (torch.Tensor, torch.Tensor)
        :return: pair of tensors of size (b x d) or (d), where
            b is the batch size and d is the dimensionality of the random
            variable. The first (second) Tensor is the lower (upper) end of
            the confidence region.
            """
           
            return self.stddev

x_range = [0,1]

x_true = torch.linspace(*x_range,100)

def f_true(x_true, noise_level = 0):
    noise_disturbance = torch.randn(x_true.shape[0]) * noise_level
    y_true =  (torch.exp(-80*(x_true-0.2)**2) + 0.6 *torch.exp(-10*(x_true - 0.6)**2))*1.5 + noise_disturbance 
    return y_true

def update_training(x_train,x_next):

    x_train = torch.cat((x_train, x_next),0)
        
        #x_train = torch.tensor(torch.linspace(0,1,4))

    y_train = f_true(x_true=x_train)

    return x_train,y_train


def build_noisy_GP(x_train,y_train,noise_level)  : 

    """
    noise_level (scalar, as float)
    """

    noise_level = noise_level
    #noise = noise_level*torch.ones(x_train.shape[0])
    mll = gpytorch.likelihoods.FixedNoiseGaussianLikelihood(noise=torch.tensor([noise_level]))
    # build the model
    gpmodel = ExactGPModel(x_train,y_train,mll)


    # show the values

    #gpmodel.state_dict()

    gpmodel.train()
    mll.train()


    return x_train,y_train,gpmodel,mll

def optimize_GP(x_train,y_train,gpmodel,mll):

   
    # Use the adam optimizer
    optimizer = torch.optim.Adam(gpmodel.parameters(),lr=0.1)

    # "Loss" for GPs - the marginal log likelihood
    loss = gpytorch.mlls.ExactMarginalLogLikelihood(mll,gpmodel)

    n_iter = 200

    for i in range(n_iter) :
        optimizer.zero_grad()

        output = gpmodel(x_train)
        loss_value = -loss(output, y_train)
        loss_value.backward() # back propagation



        #Print the loss, length scale and noise for 20 iterations
        print('Iter %d/%d - Loss: %.3f   lengthscale: %.3f  noise: %.3f ' % (
            i + 1, n_iter, loss_value.item(),  #iteration number
            gpmodel.covar_module.base_kernel.lengthscale.item(), #length scale
            gpmodel.likelihood.noise.item() #noise
                )) 
        optimizer.step()

    return gpmodel, mll, y_train

def GP_eval(gpmodel,mll, x_test = torch.linspace(*x_range, 100) ):

     """   
     Args:
        gpmodel (GPModel): The first parameter.
        mll (a likelihood): The second parameter.

    Returns:
        observed_pred: use self.mean or self.stddev to retrieve the mean or the standard deviation respectively
        
     """

     gpmodel.eval()
     mll.eval()

     with torch.no_grad(), gpytorch.settings.fast_pred_var():
          x_test = x_test
          observed_pred = mll(gpmodel(x_test)) 

     return observed_pred, x_test


def expected_improvement(mean, sigma, y_train, x_test, kappa=0):
    """
    Calculates the expected improvement for the current GP:
    Inputs: 
    mean (torch.tensor) m x n
    standard deviation (torch.tensor) m x n
    x_train (torch.tensor): necessary to obtain current max value
    x_test (torch.tensor): necessary to suggest next x to be evaluated
    kappa (scalar) : weight on exploration rather than exploitation (default kappa = 1)
    """
    try: 
        f_max = y_train.numpy().max()

    except:
        f_max = y_train.max()

    mean = mean.numpy()
    sigma = sigma.numpy()

    ### strange results occur when using f_improve = max(0,f_mean - f_max), the results when allowing negative values makes more sense. 
        
    z = (mean - f_max-kappa)/sigma # transformation from individual distribution to standard gaussian distribution N(0,1)
                                    # the mean is variable, such as the std dev, and we want to calculate the area under the bell shape from f_max to infinity. 

    stdnormalDistribution = statistics.NormalDist(0,1)
    Phi_z=[]
    phi_z=[]
    for value in z:

        Phi_z.append(statistics.NormalDist.cdf(stdnormalDistribution,x=value))
        phi_z.append(statistics.NormalDist.pdf(stdnormalDistribution,value))

    expected_improvment_array = (mean - f_max -kappa ) * Phi_z + sigma *phi_z

    #print("My approach: cdf: " + str(Phi_z)+ " pdf:" + str(phi_z))

    return expected_improvment_array

def build_test_GP(x_init, noise=0):


    y_init = f_true(x_init,noise_level=noise)

    
    x_train, y_train, gpmodel, mll =build_noisy_GP(x_init,y_init,noise_level=noise)



    gpmodel, mll, y_train = optimize_GP(x_train=x_train,y_train=y_train,gpmodel = gpmodel,mll=mll)



    observed_pred, x_test = GP_eval(gpmodel,mll) # with default x_test = linspace(0,1,100)

    print("Test GP built successfully")


    return observed_pred, x_test, x_train, y_train
```

Here is a part of the output: 

```
Iter 1/200 - Loss: 1.341   lengthscale: 0.693  noise: 0.010 
Iter 2/200 - Loss: 1.220   lengthscale: 0.693  noise: 0.010 
...
Iter 199/200 - Loss: -1.370   lengthscale: 0.693  noise: 0.010 
Iter 200/200 - Loss: -1.371   lengthscale: 0.693  noise: 0.010 
```
Also the predictions, when plotting the following code, do not look promising...:( 

```
fig, ax = plt.subplots(1,1)

ax.plot(x_test,observed_pred.mean)
ax.fill_between(x_test, observed_pred.mean + observed_pred.stddev, observed_pred.mean - observed_pred.stddev, alpha = 0.2)
plt.show()
```

I imagine that my code has much to improve. Feel free to express any concerns or suggestions, I'll be very glad to improve on that :) Thank you for any help!
